### ISSUE TITLE
Support for Raymarine Seatalk Autopilots

### ISSUE THREAD

**Issue by Leijoma** [SignalK/signalk-autopilot#22]
Hi, I have my Raymarine (Autohelm )ST7000 autopilot connected to SignalK with NMEA2000 through a Seatalk/SeatalkNG converter. 

I believe the autopilots Seatalk messages is translated to NMEA2000 using program 126720. Is this the same program used to communicate with Raymarine SeatalkNG autopilots? 

/Magnus

**Comment by sbender9** [comment]
It only uses 126720 to change the target heading. It uses 126208 for everything else.

I think it may be possible to 126720 for everything, but I don't have the info on the "key codes"

```
const keys_code = {
    "+1":      "07,f8",
    "+10":     "08,f7",
    "-1":      "05,fa",
    "-10":     "06,f9",
    "-1-10":   "21,de",
    "+1+10":   "22,dd"
}
```

**Comment by Leijoma** [comment]
I think i have most of the keycodes and also the data sent from the AP in this google sheet:

https://docs.google.com/spreadsheets/d/1R8G4HYU_DrDnAW7iDM7zKXM4d4ZkJbbpecdcOwOxfHg/edit?usp=sharing

I will take a look at your code and see if i can modify it to work. 

/Magnus

**Comment by remarkov** [comment]
Hi @Leijoma,
Did you get any progress so far? I am also using Raymarine ST/STng converter and I can see the actual data of the AP but so far did not get any luck with altering AP settings from the plugin.

**Comment by godind** [comment]
It's our pleasure to help! We have a big update ready but we need help testing first and need help with beta testers.

**Comment by remarkov** [comment]
Please let me know if I can help somehow. Probably this thread is not a proper place to continue though.

**Comment by ignamendi21** [comment]
Hello, I have a question about signalk-autopilot for Seatalk1 devices... Maybe is exactly what @godind explained...
I currently have Seatalk 1 devices (sensor and autopilot). I have implemented a Seatalk1-to-SeatalkNG converter, so I have a Nmea2000 B&G screen where I see my data. 
Having said that, I have connected a signalk server using the Nmea2000 net created after the converter. Everything works fine, I can see all data in signalk.
I read about the signalk-autopilot plugin. 
My question is: Can I control my seatalk1 autopilot with my signalk that is wire in the SeatalkNG (nmea2000) net? Will the Sealtalk1-to-SeatlkNG converter, translate back the signalk-autopilot orders? How do I configure the autopilot plugin settings?

**Comment by godind** [comment]
Hi,

Yes it should work. Well, you'll be the second person to test it. The converter will send the N2k message to contrôle head.

**Comment by remarkov** [comment]
Yes, I confirm it works perfectly well both ways, i.e. I can read the current autopilot settings in SignalK and control then from the app. Unfortunately I'm not on the boat now so can't provide any details regarding the settings, but AFAIR I only had to select the type of the autopilot and that's it. Just be sure to have the latest version.

**Comment by godind** [comment]
Great!  I have a version with minor updates on hold. Not sure I can release it but if you find issues or improvements please open here.

Enjoy

**Comment by ignamendi21** [comment]
Hello,
I tested and it work something. Currently it only work the +1, -1, +1, -10 buttons. So the autopilot need to be in "auto"... The rest of the buttons does nothing. I can see log when I press "auto", "standby","etc..." but my autopilot does nothing. I can not change to "auto", to "standby"....  The only that works is the +1, -1, +1, -10 buttons. : (

anyone faced this?

**Comment by ignamendi21** [comment]
This is exactly what I am experimenting https://github.com/SignalK/signalk-autopilot/issues/27#issue-1822534551

**Comment by ignamendi21** [comment]
> Yes, I confirm it works perfectly well both ways, i.e. I can read the current autopilot settings in SignalK and control then from the app. Unfortunately I'm not on the boat now so can't provide any details regarding the settings, but AFAIR I only had to select the type of the autopilot and that's it. Just be sure to have the latest version.

What do you mean with the "type of the autopilot"?

**Comment by ignamendi21** [comment]
![image](https://github.com/SignalK/signalk-autopilot/assets/89136198/0c6eb41b-6702-457b-ade0-1f0c00340701)
The only thing I changed from defaults values is the autoilot type, the other fields I  left them dafaults. Do I have to change their values?

**Comment by godind** [comment]
That should fine. What's your controler model?

**Comment by godind** [comment]
The Raymarine SmartPilot to STNG converter autopilot type only supports SmartPilot AP and ST model control head. The SPX AP models are n2k compatible, from what I read, and should  not use a converter.

**Comment by ignamendi21** [comment]
Thanks for answering @godind ,
My AP is SmartPilot, seatalk
As I explained, my issue is that only the button -1,+1,-10,+10 are working. Anyone with SmartPilot got "auto", "standby", etc buttons working?

I have actually tried, instead of seting "autopilot type" the "Raymarine autopilot -< Seatlk converter" , to set it in "Nmea2000". Iftdoes exactly the same thing. It Is working with the +1,+10... buttons only.

**Comment by ignamendi21** [comment]
I mean, I dont understand what can be the root of the issue. If the +1,+10,-1,-10 buttons work fine, why the "auto", "standbye" wouldnt work?? : (

**Comment by godind** [comment]
Can you share the plug-in version, AP & control head model and the app you are using to operate (the plug-in UI, KIP or something else) please. If you are using KIP's AP widget, please share the version of KIP.

**Comment by godind** [comment]
I forgot to ask, have you updated to a recent STNG Converter firmware? They have updated PNG support of the converter a few times. That could maybe be the problem.

**Comment by godind** [comment]
Hard to say at this point.

- It could be the that part of the PNG messages emitted from the app is not passing the converter or not supported by the the control head.
- Or you have different hardware/firmware versions

**Comment by ignamendi21** [comment]
I use the signalk-autopilot plugin interface. Which is updated to the 1.4.0 version
I dont know what you mean about the converter firmeware. But I bought the seatalk1 to seatalkNG converter 4 years ago.
I have SmartPilot

**Comment by godind** [comment]
The STng-Seatalk Converter firmware (the code that runs on the Raymarine device to convert n2k to seatalk messages back and forth) is probably too old but we can't tell if this is really the problem before it's updated. You could read all the firmwares update notes for every version released since the one you have running to find if they added AP PNG message support, but better to get the latest to be certain.

It might go away once updated. See the Raymarine procedure to update your n2k device firmwares.

Hope this helps

**Comment by godind** [comment]
What control head model do you have: ST6001, ST6002, ST7001...?

What SmartPilot model do you have: S1, S2, S3...?

**Comment by ignamendi21** [comment]
@godind  My Smarpilot is D40 and my displays are ST6001

**Comment by Leijoma** [comment]
I just did some tests with the latest version installed from Signalk-server.

I am using:
ST1-STngConcerter
Autohelm type 1 AP
ST7000 control

Everything seems to work as it should.
Red error message when I tried to tack in auto mode. "not in wind vane
mode", but this actually works on the st7000 panel.

In plugin config I chose Smartpilot - seatalk stng converter

Plugin identified converter automatically.



Den mån 25 sep. 2023 09:15ignamendi21 ***@***.***> skrev:

**Comment by godind** [comment]
Hi,

I don't think there was ever a D40 SmartPilot pilot dut I can be wrong. Note that a Autohelm type 1 is a drive unit (the motor that moves the rudder), not a course computer.

Here is bit of info to help you and us identify the problem:

In general and Autopilot Package (AP) is composed of:
- a course computer 
- a compas
- a drive unit
- a control head

Signal K does not support all model APs and all models, yet. We will need the proper device models to see if it should work or if we can add it by adding different setups.

Your contrôle head is a ST7000, but we need to know the course computer model as well. It should be "somewhere" inside your boat usually in a hidden compartment. See https://www.ebay.ca/itm/126081846996?chn=ps&norover=1&mkevt=1&mkrid=706-89093-2056-0&mkcid=2&mkscid=101&itemid=126081846996&targetid=1657530083670&device=m&mktype=pla&googleloc=1002589&poi=&campaignid=17288881879&mkgroupid=135489415103&rlsatarget=pla-1657530083670&abcId=9300870&merchantid=10252592&gbraid=0AAAAAD00iICAwlT454Uqz8yPOPHxhrUDm&gclid=EAIaIQobChMImsW7zeXFgQMVwVFyCh1gygDfEAQYASABEgLcsPD_BwE

For an example. Send us pictures of your Course Computer and Control Head. 

Thanks

**Comment by Leijoma** [comment]
Yes. Realized my mistake after sending the email. The autopilot computer is
a Autohelm type 100.

And to confirm. My test was a success with this AP except for the tack
button not working in auto mode.

Den mån 25 sep. 2023 14:48David G ***@***.***> skrev:

**Comment by SCarns** [comment]
Unfortunately, mine worked perfectly last summer across the STNG converter. I've updated SK through a couple of versions and have the newest AP 1.7 PI installed and I am now experiencing these issues as well. I went back to v1.3 to see if it worked, but same issues. I wish I knew what changed in SK to help sort this out.

**Comment by macjl** [comment]
Hello,

I just installed a St-Stng converter, and I’m facing the same issue as @ignamendi21. In KIP, I can set +1, -1, +10, -10, but can’t change mode or disengage the Autopilot.

The Autopilot is a SmartPilot S2. The firmware of the st-stgn converter is 2.0.8.

What can I do to collect data to understand the issue ?
